src/hip_hcc.cpp

```diff
@@ -2467,11 +2467,17 @@ hipError_t hipHccGetAcceleratorView(hipStream_t stream, hc::accelerator_view** a
 namespace hip_impl {
     std::vector<hsa_agent_t> all_hsa_agents() {
         std::vector<hsa_agent_t> r{};
-        for (auto&& acc : hc::accelerator::get_all()) {
+        auto accelerators = hc::accelerator::get_all();
+        for (int i = 0; i < accelerators.size(); i++) {
+            auto&& acc = accelerators[i];
             const auto agent = acc.get_hsa_agent();
 
             if (!agent || !acc.is_hsa_accelerator()) continue;
 
+            // If device is not in visible devices list, ignore
+            if (std::find(g_hip_visible_devices.begin(), g_hip_visible_devices.end(), (i - 1)) ==
```

Comment:
What happens if accelerators[0] happens to be a HSA accelerator? What's the value of i - 1 in that case?

Reply:
This code is duplicated from:

https://github.com/ROCm-Developer-Tools/HIP/blob/30b5c02ec48c04a4221385ad83a81afb2084b999/src/hip_hcc.cpp#L1410-L1422

Admittedly, this violates the Don't Repeat Yourself principle.  The assumption appears to be that accelerators[0] will be the CPU device and/or skipped.  If it made it as far as the std::find(), it would be looking for a visible ID of -1, which will never be in the visible device request list since it is out of range in:

https://github.com/ROCm-Developer-Tools/HIP/blob/30b5c02ec48c04a4221385ad83a81afb2084b999/src/hip_hcc.cpp#L1393-L1400

Would you suggest creating a static inline function that returns the visible device list to be used in these two places?  The logic in such a function would be updated to not assume anything, but rather it should iterate over the returned accelerators and enumerate only the GPUs, starting with an enumerated index of 0.  Something like the following?

```C++
static inline std::vector<hsa_agent_t> visible_hsa_agents() {
    std::vector<hsa_agent_t> r{};
    int accelerator_index = 0;
    for (auto&& acc : hc::accelerator::get_all()) {
        const auto agent = acc.get_hsa_agent();

        if (!agent || !acc.is_hsa_accelerator()) continue;

        // We know agent is an accelerator.
        // If device is not in visible devices list, ignore.
        // Post increment accelerator_index.
        if (std::find(g_hip_visible_devices.begin(), g_hip_visible_devices.end(), accelerator_index++) ==
                g_hip_visible_devices.end()) continue;

        r.emplace_back(*static_cast<hsa_agent_t*>(agent));
    }

    return r;
}
```

Reply:
Strictly from the standpoint of "I'd like to understand this code months / years later" the above would be highly preferable IMHO (even though it still pre-assumes that there's a bijection from index in the vector returned by get_all() and visibility info as reflected in g_hip_visible_devices, which induces implicit coupling at a distance). 

Separately, it's unclear to me if the access to g_hip_visible_devices is safe - could a situation arise in which someone is adjusting visibility in a separate thread, concurrently with this invocation of visible_hsa_agents? Or is visibility constant per process?